### PR TITLE
rustc_metadata: minor tidying

### PR DESCRIPTION
**Pull request by tamird**
Cleaned up some code while investigating #131720.

See individual commits.

### CODE REVIEW

File: compiler/rustc_metadata/src/locator.rs [hunk 1]
```diff
@@ -721,28 +721,33 @@ impl<'a> CrateLocator<'a> {
                 return Err(CrateError::ExternLocationNotFile(self.crate_name, loc.clone()));
             };
 
-            if file.starts_with("lib") && (file.ends_with(".rlib") || file.ends_with(".rmeta"))
-                || file.starts_with(self.target.dll_prefix.as_ref())
-                    && file.ends_with(self.target.dll_suffix.as_ref())
-            {
+            match file.starts_with("lib").then_some(()).and_then(|()| {
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/locator.rs:724]: ? isn't this just `.then`

Comment (tamird) [reply compiler/rustc_metadata/src/locator.rs:724]: No, because the closure still needs to return `Option<T>` rather than `T`.

But this can be `.then(...).flatten()`. Making that change causes rustfmt to increase indentation, so I've left it as is.

Comment (jieyouxu) [reply compiler/rustc_metadata/src/locator.rs:724]: To be honest I find this `.then_some(()).and_then(|()| {` very hard to read. Can't we just do something straightforward like https://github.com/jieyouxu/rust/commit/6eba971bde7c2a05448e0073f0da100a145c4568?

Comment (tamird) [reply compiler/rustc_metadata/src/locator.rs]: Thanks for the suggestion. I was trying to avoid repeating conditions (your code repeats `file_name.starts_with("lib")` twice).

I replaced this with a more imperative closure which should be easier to understand. Can you take a look?

Comment (jieyouxu) [reply compiler/rustc_metadata/src/locator.rs]: Frankly, I don't think it's worth avoiding the trivial repetitions, I still find this hard to follow than just doing the simple thing. I let other compiler reviewers decide what they find more readable.

Comment (tamird) [reply compiler/rustc_metadata/src/locator.rs]: Would you mind removing the waiting-on-author label?

Comment (jieyouxu) [reply compiler/rustc_metadata/src/locator.rs]: > Would you mind removing the waiting-on-author label?

You can write `@rustbot ready` to indicate that.

File: compiler/rustc_metadata/src/locator.rs [hunk 2]
```diff
@@ -706,54 +709,57 @@ impl<'a> CrateLocator<'a> {
...
+                None
+            })() {
+                Some(dst) => {
+                    dst.insert(loc_canon.clone(), PathKind::ExternFlag);
+                }
+                None => {
+                    self.crate_rejections
+                        .via_filename
+                        .push(CrateMismatch { path: loc_orig.clone(), got: String::new() });
                 }
```
Comment (lcnr) [inline compiler/rustc_metadata/src/locator.rs:757]: I am personally not a fan of matching on a closure, maybe something like this?
```rust
            'accept_crate: {
                if file.starts_with("lib") {
                    if file.ends_with(".rlib") {
                        rlibs.insert(loc_canon.clone(), PathKind::ExternFlag);
                        break 'accept_crate;
                    }
                    if file.ends_with(".rmeta") {
                        rmetas.insert(loc_canon.clone(), PathKind::ExternFlag);
                        break 'accept_crate;
                    }
                }
                let dll_prefix = self.target.dll_prefix.as_ref();
                let dll_suffix = self.target.dll_suffix.as_ref();
                if file.starts_with(dll_prefix) && file.ends_with(dll_suffix) {
                     dylibs.insert(loc_canon.clone(), PathKind::ExternFlag);
                     break 'accept_crate;
                }
                
                self.crate_rejections
                        .via_filename
                        .push(CrateMismatch { path: loc_orig.clone(), got: String::new() });
           };
```

Comment (tamird) [reply compiler/rustc_metadata/src/locator.rs]: That's fragile because removing any `break` would produce incorrect logic. I've assigned the result to a local variable, do you like that better?

**Comment by lcnr** [comment]
good enough :grin: 

@bors r+ rollup
